addon/components/bs-modal.hbs

```diff
@@ -1,14 +1,15 @@
 {{did-insert this.handleVisibilityChanges}}
 {{did-update this.handleVisibilityChanges @open}}
-{{#if this.inDom}}
+
+{{#if this.isModalShown}}
   {{#let (component (bs-default @dialogComponent (component "bs-modal/dialog"))
     onClose=this.close
     fade=this._fade
-    showModal=this.showModal
+    showModal=this.isModalShown
     keyboard=this.keyboard
     size=@size
     backdropClose=this.backdropClose
-    inDom=this.inDom
+    inDom=this.isModalShown
```

Comment:
This seems suspicious, as `@showModal` and `@inDom` receive the same value, thus seem redundant. But I don't think they are. IIRC having those separate args might be related to having to put the modal into DOM first, and then add the `in` class to trigger the CSS transition. Doing it at the same time might prevent the transition to happen correctly. Which is likely not covered by tests.

Reply:
Good catch. In current implementation `show` method sets `inDom` to `true` immediately while `showModal` is set to `true` _after_ the backdrop is shown in `afterRender` runloop.

Hard to follow the existing code. But if I did not missed something it's simplified like this:

```js
async show() {
  this.inDom = true;
  
  await this.showBackdrop();

  schedule('afterRender', () => {
    this.showModal = true;
  });
}
```

Will try to manual check if I see a visual difference before and after. If so I will try to add test coverage for it.

Reply:
Looking at the `<BsModal::Dialog>` component the `@showModal` and `@inDom` arguments are consumed in https://github.com/kaliber5/ember-bootstrap/blob/6aeead2d17604afbdccc4295a4dffba621771239/addon/components/bs-modal/dialog.hbs#L5 and https://github.com/kaliber5/ember-bootstrap/blob/6aeead2d17604afbdccc4295a4dffba621771239/addon/components/bs-modal/dialog.hbs#L11.

I only consider Bootstrap 4 for now.

The `@showModal` argument controls if `.show` CSS class is applied. The `@inDom` argument controls if `.d-block` CSS class and `display: block` CSS is applied. So the question is `.show` CSS class and `display: block` should be applied at the same time or not.

Bootstrap applies both the `.show` CSS class and sets `display: block` in the same `_showElement` method: https://github.com/twbs/bootstrap/blob/20cfbdff79e19e77c24596cdca59c6694283e242/js/src/modal.js#L231-L254 So there doesn't seem to be timing difference for showing the modal.

But for hiding the modal the `.show` CSS class is removed _before_ the animation starts and `display: none` is set _after_ the animation is over. https://github.com/twbs/bootstrap/blob/20cfbdff79e19e77c24596cdca59c6694283e242/js/src/modal.js#L178 and https://github.com/twbs/bootstrap/blob/20cfbdff79e19e77c24596cdca59c6694283e242/js/src/modal.js#L316

Unless I missed something `@showModal` should be true for `opening` and `open` state, while `@inDom` should be true for `opening`, `open` and `closing` states.

Applying both the `.d-block` CSS class and `display: block` if `@inDom` is `true` seems to be redundant. See [live demo in Bootstrap documentation](https://getbootstrap.com/docs/4.6/components/modal/#live-demo) seem to have only `display: block` CSS property. But I guess for FastBoot support we need to apply the CSS class as well. At least as long as ember-style-modifier does not have support FastBoot yet.

Reply:
I don't get why `@inDom` argument is passed at all to `<BsModal::Dialog>`. Also in current implementation the component is teared down if `this.inDom` is `false`: https://github.com/kaliber5/ember-bootstrap/blob/6aeead2d17604afbdccc4295a4dffba621771239/addon/components/bs-modal.hbs#L3 It looks as if we are doing this:

```hbs
{{#if this.inDom}}
  <Foo @bar={{this.inDom}} />
{{/if}}
```

This seems to be the same as

```
{{#if this.inDom}}
  <Foo @bar={{true}} />
{{/if}}
```

because the code path is only triggered for `this.inDom === true`. :confused:

Reply:
Animation timing should now be the same as before. At least for `.modal` itself. Haven't added test coverage for backdrop timing yet.

Reply:
> Bootstrap applies both the .show CSS class and sets display: block in the same _showElement method:

I think the difference here between plain BS and our components is that by having rich interactive components instead of static markup created by the user and some JS on top, we are able (and actually doing that) to render the modal's DOM only when needed, hence the `{{#if this.inDom}}`. But in this case we must make sure that first the DOM is created, and *after* that we add the `show` class, otherwise the CSS transitions (triggered by adding `show`) would not apply. That's why we also need these two different but similar states: `inDom` and `showModal`. Does that make sense?

> Applying both the .d-block CSS class and display: block if @inDom is true seems to be redundant. 

> I don't get why @inDom argument is passed at all to <BsModal::Dialog>

Seems you are right about these points. I can only explain it that these things were messed up during the course of refactorings, and indeed the state management of this component is a bit confusing.